Metaschema XSLT Inspector

wendellpiez opened the pull request:
# Committer Notes

Working towards #72

### All Submissions:

- [ ] Have you followed the guidelines in our [Contributing](https://github.com/usnistgov/metaschema-xslt/blob/master/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/usnistgov/metaschema-xslt/pulls) for the same update/change?
- [ ] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [ ] Do all automated CI/CD checks pass?

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you included examples of how to use your new feature(s)?
- [ ] Have you updated all [website](https://pages.nist.gov/metaschema) and readme documentation affected by the changes you made? Changes to the website can be made in the website/content directory of your branch.

wendellpiez committed:
First rough cut

wendellpiez committed:
Renamed 'inspector' to 'InspectorXSLT' for clarity.

wendellpiez committed:
Starting to frame out Inspector XSLT production and testing

wendellpiez committed:
Adding first sets of unit tests; more infrastructure too.

wendellpiez committed:
Further work on patterns; simplifying XSLT entry for testability.

wendellpiez committed:
Moving patterns into separate patterns xspec file

wendellpiez committed:
Updates with more unit tests - now testing templates in application; new invalid sample; Schematron for XSLT line number alignment assurance.

wendellpiez committed:
Found and repaired missing feature in requirements testing (`choice`), with metaschema mod illustrating a choice; two new samples.

wendellpiez committed:
Another use case tested

wendellpiez committed:
Forgot to unmark pending - another test now runs

wendellpiez committed:
Now with some post-processing demonstrated - HTML results

wendellpiez committed:
Further enhancement to demo html production

wendellpiez committed:
Continuing implementation; testing; XSLT runtime interface

wendellpiez committed:
More implementation and testing including datatype (lexical) testing; refining tests and reports - wip

wendellpiez committed:
Top-level scripting generating/refreshing Inspector XSLT

wendellpiez committed:
Improved transparency in datatype lexical checking

wendellpiez committed:
Incremental improvements to tests and docs

wendellpiez committed:
More refinements and aesthetic enhancements to reporting

wendellpiez committed:
Further refinements to runtimes, scripting, docs

wendellpiez committed:
Initial cut at some constraints when target=".", with some tests

wendellpiez committed:
More cleanup; refinement; tests for allowed-values; bug repairs

wendellpiez committed:
More various improvements to tests; reporting; unit tests; readme.md

wendellpiez committed:
Woops, forgot to save newest readme with last commit.

wendellpiez committed:
Some cleanup; readme work

wendellpiez committed:
Constraints cascade now framed out

wendellpiez committed:
Further with constraints implementation with unit tests

wendellpiez committed:
More stabilization - cross-listing tests with rule types

wendellpiez committed:
More progress implementating and testing constraints - expect/@test

wendellpiez committed:
Updating tests and valid instances

wendellpiez committed:
has-cardinality constraint implementation and testing; alignment

wendellpiez committed:
Refining messages

wendellpiez committed:
XSpec exercising constraints testing somehow wasn't added earlier

wendellpiez committed:
Touchups

wendellpiez committed:
Added new 'compressed Markdown' plain ASCII result variant with mode=compressed

wendellpiez committed:
Touchups; mode=compressed is now mode=compact

wendellpiez committed:
Further stabilization of interfaces; test updates

wendellpiez committed:
Updated unit tests etc with is-unique implementation; more readme work

wendellpiez committed:
Adjusting scripting with help

wendellpiez committed:
Updates mostly polishing; cl interfaces (new stylesheet parameter for output)

wendellpiez committed:
Added support for simpleType extensions in datatypes; updated test Computer Model metaschema aligning with metaschema repo model; changes accordingly to unit tests - updates still to be made to constraints unit testing before completing the feature set

wendellpiez committed:
Now with an implementation of index-has-key - while many tests are turned off pending update

wendellpiez committed:
Valid samples now valid again, including to defined constraints.

wendellpiez committed:
Updating readme.md, XSpec

wendellpiez committed:
Filling out tests; support for scenario numbering in XSpec

wendellpiez committed:
More updates including a start at XSLT production unit testing

wendellpiez committed:
New version in alpha2, stable so-far version in alpha1

wendellpiez committed:
Updates, adjustments and rearrangements -- XSpecs currently ALL GREEN

wendellpiez committed:
Some more renaming and ordering; readme updates

wendellpiez committed:
Cleanup and mopup; simplifying interface; readme

wendellpiez commented:
See wiki page here: https://github.com/usnistgov/metaschema-xslt/wiki/InspectorXSLT

Discussion board to come (watch for link).

wendellpiez committed:
Renamed testing resources for gitignore convenience; gitignore touchup; some cleanup

aj-stein-nist commented on src/schema-gen/InspectorXSLT/TESTING.md:
```diff
@@ -0,0 +1,107 @@
+
+### "Computer Model" metaschema
+
+This suite uses an extension of AJ Stein's 'computer metaschema' model for testing, maintained here as [testing/computer_metaschema.xml](testing/computer_metaschema.xml).
```
Minor, non-blocking: technically it's the Metaschema project's model. Dave and others have made significant improvements there.

```suggestion
This suite uses an extension of 'computer metaschema' model from [examples included in the specification's repository](https://github.com/usnistgov/metaschema/blob/develop/examples/) for testing, maintained here as [testing/computer_metaschema.xml](testing/computer_metaschema.xml).
```

aj-stein-nist commented on src/schema-gen/InspectorXSLT/testing/readme.md:
```diff
@@ -0,0 +1,74 @@
+- [ ] CSX version
+   - [ ] CI/CD support for deployment from updated metaschema sources?
+- [ ] Release plans and considerations
+- [ ] Review opportunities for unit tests: reuse / pubs / pool 
```
Minor: I wouldn't block over this, but it seems better to have issues than inline TODOs in READMEs deep in the directory structure that would get lost. I would convert them to issues in the GH repo itself.

```suggestion
```

aj-stein-nist reviewed (commented):
Initial feedback, but more to follow with an approval later today or tomorrow when I experiment with it more and make sure I don't find any major problems. Looks good so far. I know we discussed README organization and separation. I would think it is better to make it in a different repo to not confuse consumers of this repo, but you are the maintainer. Not my repo! :-)

wendellpiez committed:
Renaming, cleanup and readme edits as suggested by AJ

wendellpiez committed:
Minor but significant usability tweaks in docs

wendellpiez committed:
Couple more similar improvements

wendellpiez committed:
These are getting smaller, honest.

wendellpiez committed:
Further bit of clarification

wendellpiez committed:
Adding Schematron computer-inspector harness; more edits to readme.md

wendellpiez committed:
Reporting in Schematron harness is a little less hopeless

wendellpiez committed:
Further improvement to Schematron harness / legibility

wendellpiez commented:
Closes #72

wendellpiez committed:
Update to readme

wendellpiez commented:
Holding back pending

- Support for all paths in OSCAL (cf #91 #92)
- Support for updated `allowed-values` semantics (cf https://github.com/usnistgov/metaschema/pull/413)
- Integrating PR #94

wendellpiez committed:
helping gitignore ignore XSpec results?

wendellpiez committed:
Added 'OSCAL patcher' production capability to bridge OSCAL/Metaschema 1.0 delta, with XSpec updates

wendellpiez committed:
More improvements to XSpecs esp wrt namespace declarations etc.

wendellpiez committed:
Now able to emit a functional XSLT for the OSCAL catalog metaschema

wendellpiez committed:
Repairing XSLT generation glitch

aj-stein-nist committed:
Test in GHA CI, use make for InspectorXSLT for #72

This change integrates the Makefile-based testing harness for each of
the submodules.

For the InspectorXSLT submodule in the schema-gen directory, integrate
the shell script that orchestrates Maven's installation of dependencies
and execution of XML Calabash to run XProc files. The XProc pipeline
files will run the XSpec tests in batch mode.

aj-stein-nist reviewed (approved):
@wendellpiez, I am re-reviewing and approving at the time of this message to signal I am happy with the current version of the CI additions, now with more STDOUT and STDERR logging to move us forward. If you want me to re-review after you add relevant documentation, feel free to do so, but I already reviewed the other XSLT, XSpec, XProc, and shell script changes here as needed to advance this work.

wendellpiez committed:
Adjusting Makefile; improved TESTING.md

wendellpiez committed:
Further touchups and experiment

wendellpiez commented:
@aj-stein-nist please feel free to re-review changes -- mainly in TESTING.md documentation (not perfect yet but better), plus also a couple of tweaks elsewhere.

The remaining desideratum for this Issue is to have CI on a PR report failure not only for broken runs but also for successful runs reporting test failures. I have vacillated but I think this is more or less essential.

A cheap/dirty way to do this is by inspecting the outputs for key (regex) match patterns.

And there could be better options. (Maybe our testing could produce a format that Github Actions could use?) Happy for any thoughts.

wendellpiez committed:
Adding more test files for ci/cd XSpec

wendellpiez commented:
In the current 'integration tests' (smoke tests) we have failures successfully reported by at least one XSpec -

https://github.com/usnistgov/metaschema-xslt/actions/runs/7791678616/job/21248238501?pr=73#step:5:144

We have the option of making either script or XSLT runtime error out, but it makes sense to catch this gracefully, *non*? (Unless an actual reporting-XML concept takes hold.)

wendellpiez committed:
In Saxon-based XSpec batch pipeline, an optional parameter now provides for JUnit XML test results to be produced and written.

wendellpiez committed:
Bit more touchup for CI/CD

wendellpiez committed:
First attempt at trying JUnit reports on XSpec under GHA
